### ISSUE TITLE
Make explicit what HTTP Verbs are to be supported by an XProc 3.0 processor

### ISSUE THREAD

**Issue by fsasaki** [xproc/3.0-steps#42]
The current draft of the "steps" specification says:
"The method attribute specifies the method to be used against the IRI specified by the href attribute, e.g. GET or POST (the value is not case-sensitive). If the href attribute is not absolute, it will be resolved against the base URI of the element on which it is occurs."
Two comments on that
1) It should be made explizit what HTTP Verbs are to be processed by an XProc 3.0 processor
2) The following verbs should be mandatory for a processor: GET, POST, PUT, DELETE, HEAD .
Rationale: for using XProc in a micro-services environment, having HTTP verbs available would allow to create an "XProc only" workflow, without the need to use a different (non XML) technologies just for the service communication layer.

**Comment by ndw** [comment]
We'll update the spec to say that implementations are required to document the verbs that they implement and that they SHOULD implement GET, POST, PUT, DELETE, and HEAD (for HTTP).

**Comment by Conal-Tuohy** [comment]
I am wary about explicitly listing the verbs that SHOULD be supported, because I think it would send a message to XProc implementors that they need a "white-list" of HTTP methods (and that verbs outside of that list should be rejected). But in fact `http-request` doesn't need a white-list of HTTP methods, any more than it needs a white-list of supported HTTP headers. The set of HTTP verbs is arbitrarily extensible, and `p:http-request` only needs to make the HTTP request as specified by the `c:request` document; the HTTP verb specified there is just a token in the HTTP message, which the `http-request` step only needs to pass on to the HTTP server to interpret. Of course if `p:http-request` recognises a verb it may merit some special treatment (such as e.g. caching of GET requests), but if the verb is unknown then that's no reason for reject it. As someone who has had to use PATCH for dealing with a metadata repository, and had to _ahem_ patch XMLCalabash to add support, I think explicitly enumerating the verbs which SHOULD be "supported" might have the practical effect of unnecessarily limiting support for the less common HTTP verbs (such as OPTIONS and PATCH and the zillion other verbs used in WebDAV) and any new or custom verbs. My preference would be to say that the `p:http-request` step should support _any and all_ HTTP request methods.

**Comment by fsasaki** [comment]
I am fine by the resolution proposed at https://github.com/xproc/3.0-steps/issues/42#issuecomment-500448989 - thanks, Norm.

**Comment by Conal-Tuohy** [comment]
The PATCH method is another one which does get some use. I have used it myself last year with the web API of Symplectics Elements (a metadata repository), and I've also used it with Fedora Repository, which implements the Linked Data Protocol, in which PATCH support is mandatory. It's also part of the SPARQL Graph Store Protocol. If there's to be a list of methods (which I still think is a mistaken idea), then I would really like PATCH to be on it.
